### DISCUSSION TITLE
How to convert/change from LAMBDA to FILTERS?

### DISCUSSION THREAD

**Discussion by Photelegy** [dbuezas/lovelace-plotly-graph-card#234] [Q&A] [answered]
Hi everyone,
I know that there are now "filters:" which should be easier to use than "lambda:" (https://github.com/dbuezas/lovelace-plotly-graph-card#filters).
But I don't really understand how they work.
How can I replace the following "lambda" with "filters"?
(This lambda takes the graph and smoothes it (e.g. 10 seconds) so that short spikes get flattend.)

```
lambda: |-
  (ys, xs) => {
    const SECONDES = 10               // <-- change here
    const window_size_ms = 1000*SECONDES 
    const result = {x:[], y:[]}
    let x_next = +xs[0] + window_size_ms
    let y_accum = 0
    let y_count = 0
    for (let i = 0; i < xs.length; i++) {
      const x = +xs[i]
      const y = +ys[i]
      if (x>x_next){
        result.y.push(Math.round(y_accum/y_count * 10) / 10)
        result.x.push(new Date(x_next))
        y_accum = 0
        y_count = 0
        x_next = +x + window_size_ms
      }
      if (y){
        // ignore unavailable/unknown state points
        y_accum += y
        y_count ++
      }
    } 
    return result;
  }
```
![image](https://user-images.githubusercontent.com/61714583/216773914-559e17f4-af53-46bc-ba42-0510d132ad1e.png)

I searched e.g. for "rounding", "flattening" (or similar) but couldn't find a filter.

Thank you all so much!
Kind Regards
Photelegy

**Comment by dbuezas** [comment, marked as answer]
This is the one you need: sliding_window_moving_average

You were almost there, searching the readme for "smoothing" would have gotten there.

The filters are used like this:
```yaml
entities:
  - entity: sensor.temperature
    filters:
      - sliding_window_moving_average
      - here you can put more filters. They are applied in sequence
```

Look in the readme for the extra options it has.

**Reply by Photelegy** [reply]
Thank you very much for your answer.
Sorry, but I don't really understand the filter "sliding_window_moving_average".
So I can use this filter at the same places I used "lambda" (directly under "entity:" or under "defaults: -> entity:")?

So I tried the following:
```
filters:
  - sliding_window_moving_average:
      window_size: 10
      extended: false
      centered: true
```
But I don't understand how to tell it in this filter how much smoothing (as time) I want.
In the "lambda" (in my original post-example) I could say "SECONDES = 10", so the smoothing was over 10s (or change it to more or less).
Is there a parameter in the filter to tell it the time duration for the smoothing?

Thank you very much for your help!

PS:
And "filters:" should be moved in by 2 spaces, right? (If I'm right, this is a minor flaw in the documentation ✌😉)
![image](https://user-images.githubusercontent.com/61714583/217509464-1c90adb3-eb04-4fa3-a8d4-b3a75044f4b6.png)

**Reply by dbuezas** [reply]
Oh, thanks for pointing that out! Do you want to try to fix it yourself?

**Reply by dbuezas** [reply]
You can do it from the browser. While seeing the readme file, at the top right there is a pen icon:
![image](https://user-images.githubusercontent.com/777196/217649576-c8909137-d551-4a6f-a269-bfc2340c71a7.png)

Then you can add the missing spaces and open a pull request, like this:
![image](https://user-images.githubusercontent.com/777196/217649721-6e5a8876-f421-41c7-8099-c9734ba06370.png)

After that I can merge it. You'll appear as a contributor, and now that you know how, maybe decide to also fix readmes in other open source projects too :)

**Reply by dbuezas** [reply]
Regarding your question about the window size, the number represents the number of datapoints, not time lengths.
Supporting time windows would actually be an interesting feature to add (something like `window_size: 5h`). Feel free to make a feature request and I'll give it a try when I have some free time

**Reply by Photelegy** [reply]
Thanks for your answer.
I'd like to fix the readme and contribute in this way.
But somehow I don't see this pen icon on my screen. Am I right that it should be where I made the red circle?
![image](https://user-images.githubusercontent.com/61714583/217812567-79dfe493-6779-484d-8716-2c05cf029b8b.png)
Do I need to activate something or do something before I can edit the readme?

Kind Regards
Photelegy

**Reply by dbuezas** [reply]
That's odd. What happens if you click on the readme.md file and try from there?

**Comment by Photelegy** [comment]
About the feature request.
Is this done right under "Issues", "Labels: enhancement"?
https://github.com/dbuezas/lovelace-plotly-graph-card/issues/248#issue-1577841373

**Reply by dbuezas** [reply]
Great, thank you. I followed up inside that ticket.

Did you try clicking on the readme file before trying the edit? Did that work?
